### PR TITLE
Add types annotations

### PR DESCRIPTION
**Pull request by destag**
This is part of #120 and #140

### CODE REVIEW

File: requirements_dev.txt [hunk 1]
```diff
@@ -6,4 +6,5 @@ codacy-coverage==1.3.11
 codecov==2.0.15
 coverage==4.5.3
 coveralls==1.8.1
-pytest-cov==2.7.1
\ No newline at end of file
+pytest-cov==2.7.1
+mypy==0.730
```
Comment (toonarmycaptain) [inline requirements_dev.txt]: Please alphabetize these.

**Comment by toonarmycaptain** [comment]
Thankyou for your contribution. :) 
I've asked for a few changes/explanations, but mostly it looks great. 

As soon as you reorder the requirements_dev.txt I'll merge it, and then CI can run.

**Comment by destag** [comment]
Thank you for your comments. I've changed global variables default values because mypy aren't doing well detecting if variable has changed type. But we can suppress some warnings.

File: dionysus_app/data_folder.py [hunk 2]
```diff
@@ -40,8 +40,8 @@ def generate_rel_path(path):
         :return: Path object.
         """
         if not path:
-            return Path.cwd().as_uri()
+            return Path.cwd()
 
-        path = path.split('/')
-        path.insert(0, ROOT_DIR)
-        return Path(*path).resolve()
+        path_parts = path.split('/')
```
[the remark this answers is not in the document]
Comment (toonarmycaptain) [reply dionysus_app/data_folder.py]: Wait, why did you remove `.as_uri()`? I didn't notice this change earlier, but this is breaking a test, as the test uses the iterability of the `.as_uri()` strings to ensure the cwd path and the rel path are in the path strings for the app folders. 

We *shouldn't* need this functionality outside of testing. I'll submit a patch to fix the tests to use Path objects returned by the function.

File: definitions.py [hunk 3]
```diff
@@ -1,10 +1,10 @@
 import os
 
 from pathlib import Path
-from typing import List, Optional
+from typing import List, Optional, Iterable
 
 ROOT_DIR = os.path.dirname(os.path.abspath(__file__))  # Global root directory.
 
-REGISTRY: Optional[List] = None
+REGISTRY: List[str] = []
```
[the remark this answers is not in the document]
Comment (toonarmycaptain) [reply definitions.py]: NB This can be `REGISTRY: Optional[List[str]] = None` as the registry is a list of strings.

**Comment by toonarmycaptain** [comment]
> Thank you for your comments. I've changed global variables default values because mypy aren't doing well detecting if variable has changed type. But we can suppress some warnings.

I think most of the issues are where mypy is running on test code. Excepting test code from the mypy check should fix most of these. 
If there's a way to suppress the specific errors where mypy is noting that some of the variables in definitions.py are initialised with `None`, but will not be used at runtime without an appropriate type being assigned to them, that would be good. I'm leery of suppressing warnings arbitrarily. 

This will also clear up BetterCodeHub complaining about code in comments, as they won't be needed in the tests. Some tests deliberately pass unexpected types in order to check that the appropriate behaviour (handling, exception etc) occurs.

File: .travis.yml [hunk 4]
```diff
@@ -37,5 +37,6 @@ script:
     - coverage xml -i
     - python-codacy-coverage -r coverage.xml
     - codecov
+    - mypy --ignore-missing-imports --show-error-codes .
```
Comment (toonarmycaptain) [inline .travis.yml]: I'm going to add `pytest-mypy` to requirements_dev, and run the mypy check through pytest along with the coverage check. I'll also add a mypy.ini skipping import errors.

**Comment by toonarmycaptain** [comment]
@all-contributors please add @destag for infra, doc

**Pull request merged**